### PR TITLE
Keepalive fixes

### PR DESCRIPTION
**Pull request by steverweber**
keepalive code now uses exit codes and can be triggered from under a thread.
#26888 #12540

### CODE REVIEW

**Comment by steverweber** [comment]
the keepalive code was broken after some good code cleanup in the minoin.py.  Keepalive does not have a test case and was fragile code (it required error to be raised up the stack through multiple functions)... So I changed some code to use os._exit() so it bypasses that mess.

This is still under testing ... ill likely rebase when its all good.

**Comment by steverweber** [comment]
I think this is good to merge. I did the keepalive code in the past so I can vouch this is a good direction.  However this solution is still not the holy grail.  see: #26888 seems that salt master cant live on Azure.

**Comment by steverweber** [comment]
I don't like to let PR sit too long so please comment or merge.

This pr fixes the keepalive code that was broken after adding tornado to the transport

**Comment by cachedout** [comment]
I will take a look at this as soon as I am able to. @jacksontj will probably also want to have a look here, if he's free.

**Comment by steverweber** [comment]
fyi this PR does not fix #26888. However some good investigation has occurred there.

Also perhaps some feedback on if `pub_channel` can be used to preform the ping master.  Using a new connection to ping the master seems to always pass even tho the master can't see the minion.

thanks

File: salt/defaults/exitcodes.py [hunk 1]
```diff
@@ -34,5 +34,8 @@
 
 # The Salt specific exit codes are defined below:
 
+# keepalive exit code is a hint that the process should be restarted
+SALT_KEEPALIVE = 99
```
Comment (jacksontj) [inline salt/defaults/exitcodes.py]: This config is the keepalive interval? Or the exit code? Probably should rename the option to make it more clear:)

File: salt/minion.py [hunk 2]
```diff
@@ -1676,7 +1676,13 @@ def tune_in(self, start=True):
         ping_interval = self.opts.get('ping_interval', 0) * 60
         if ping_interval > 0:
             def ping_master():
-                self._fire_master('ping', 'minion_ping')
+                if not self._fire_master('ping', 'minion_ping'):
+                    if not self.opts.get('auth_safemode', True):
+                        log.error('** Master Ping failed. Attempting to restart minion**')
+                        delay = self.opts.get('random_reauth_delay', 5)
```
Comment (jacksontj) [inline salt/minion.py]: Do we need this? Doesn't fire_master do this backoff? (not sure TBH)

File: salt/scripts.py [hunk 3]
```diff
@@ -122,28 +112,21 @@ def salt_minion():
     # keep one minion subprocess running
     while True:
         try:
-            queue = multiprocessing.Queue()
+            process = multiprocessing.Process(target=minion_process)
+            process.start()
         except Exception:
```
Comment (jacksontj) [inline salt/scripts.py]: Probably should have a more specific exception :/

**Comment by jacksontj** [comment]
I'm still of the opinion that the process manager for the minion is overkill-- as every init variant supports something similar. I created #26929 for that issue, but this PR looks like a good set of fixes for the feature while it still exists :)

File: salt/defaults/exitcodes.py (hunk 1 above)
Comment (steverweber) [inline salt/defaults/exitcodes.py]: exit code... the comment above the line seems clear to me.

File: salt/minion.py (hunk 2 above)
Comment (steverweber) [inline salt/minion.py]: need this... would be nice to reactor another time.

**Comment by steverweber** [comment]
see:
https://github.com/saltstack/salt/pull/22313
https://github.com/saltstack/salt/issues/12540#issuecomment-72895359

**Pull request merged**